src/core/catalog.js

```diff
@@ -1286,6 +1290,7 @@ class Catalog {
       // and to ensure that we actually find the correct `Page` dict.
       for (let last = kids.length - 1; last >= 0; last--) {
         nodesToVisit.push(kids[last]);
+        this.pageDictCache.put(kids[last], xref.fetchAsync(kids[last]));
```

Comment:
This is missing some necessary validation, and it could also lead to promises being replaced unnecessary. (Finally, a local variable wouldn't hurt either.)
```suggestion
        const lastKid = kids[last];
        nodesToVisit.push(lastKid);
        
        if (lastKid instanceof Ref && !pageDictCache.has(lastKid)) {
          pageDictCache.put(lastKid, xref.fetchAsync(lastKid));
        }
```

Reply:
I'm fine with this